### PR TITLE
Add support for Mamba Basic MK3.5 Boards

### LINKED ISSUE #17989: Mamba f405 basic MK3 support
I want Ardupilot for Mamba f405 basic Mk3 board. The previous Mk2 version of this card is supported. Barometer is added in Mk3 version but no Ardupilot software. Can you help me with these with a little touch to the Mk2 software?

https://www.diatone.us/collections/basic/products/mamba-basic-f405-mk3-40a-6s-8bit-flight-controller-stack

### PR DESCRIPTION
**Pull request by tenten8401**
This adds support for the Diatone Mamba Basic MK3 line of flight controllers, and possibly more that are using MambaF405US-I2C target.

I've tested this on a quadcopter and it seems to fly fine with all uarts & i2c functional. I'm not sure whether to leave RCIN as NODMA or not because CRSF doesn't work real well without it enabled.
basic20pinout

Closes #17989

### CODE REVIEW

**Comment by FairyTalesl** [comment]
Thank you for support.  Mamba f405 mk3 board has F.port reserved for uart 5 and hardware inverter circuit is available.
The hardware inverter circuit is reserved for the uart5 pin, is there a need to set it in software? I will test it as soon as possible.

**Comment by FairyTalesl** [comment]
This helps you
https://cdn.shopify.com/s/files/1/0027/2708/4144/files/MAMBA_Basic_F405_MK3_F50.jpg?v=1616141748

**Comment by tenten8401** [comment]
@FairyTalesl I think you just use the InvertRX on UART1, or possibly the dedicated Fport pad? Give the Fport pad a try first and if that doesn't work try InvertRX on UART1
I don't have any Fport stuff to test anymore and only have the 20x20 variant so I can't test :(

**Comment by FairyTalesl** [comment]
In Betaflight only fport was working on uart5. It does not work because there is no external circuit on other pins. It would work if there was an F7 processor. I will test it as soon as possible and let you know.
This helps you 
https://oscarliang.com/setup-frsky-fport/

**Comment by tenten8401** [comment]
> In Betaflight only fport was working on uart5. It does not work because there is no external circuit on other pins. It would work if there was an F7 processor. I will test it as soon as possible and let you know.
> This helps you
> https://oscarliang.com/setup-frsky-fport/

Hmm I see. Yeah you may just have to use uart5 with fport pad then just like betaflight, and then set that to rcin protocol.

File: libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat [hunk 1]
```diff
@@ -0,0 +1,167 @@
+# hw definition file for processing by chibios_pins.py
+# Mamba Basic F405 US I2C
+# With F405 MCU, MPU6000 IMU and 7456 series OSD
+# Based on Mamba F405 from jeanphilippehell
+@ thanks to betaflight for pin information
+
+# MCU class and specific type.
+MCU STM32F4xx STM32F405xx
+
+HAL_CHIBIOS_ARCH_F405 1
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat:10]: this should not be needed

File: libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+define BOARD_RSSI_ANA_PIN 1
+
+# Setup for OSD
+define OSD_ENABLED 1
+define HAL_OSD_TYPE_DEFAULT 1
+# Font for the osd
+ROMFS_WILDCARD libraries/AP_OSD/fonts/font*.bin
+
+# To complementary channels work we define this
+#define STM32_PWM_USE_ADVANCED TRUE
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat:165]: this should be automatic now

**Review by tridge — CHANGES_REQUESTED** [review]
just a couple of small issues, thanks for doing the docs!

File: libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat (hunk 2 above)
Comment (tenten8401) [reply libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat]: These both should be good to go now

File: libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat (hunk 1 above)
Comment (tenten8401) [reply libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat]: Removed the hal_chibios part, wouldn't compile without mcu

**Comment by tenten8401** [comment]
What are your thoughts on the 64kb reserved flash? I notice storage size is 16k and bootloader is 16k  but we have 64k reserved.. This is the same on the MambaF405v2 board and presumably whatever it got copied from

File: libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat (hunk 2 above)
Comment (tenten8401) [reply libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat]: I'm not sure whether you were talking about the romfs wildcard or both, I removed both but can revert

**Comment by FairyTalesl** [comment]
Hello ı am flashed to firmware but i cant connect receiver. How to setup Serial port 1 SBUS setting ? 
Ekran Alıntısı

**Comment by tenten8401** [comment]
Can you send a picture/diagram of your wiring? Did the wiring work correctly on another flight software?

Setting to RCIN should be enough to make it work. 

Jul 13, 2021 2:28:43 PM FairyTalesl ***@***.***>:

**Comment by FairyTalesl** [comment]
İts working Betaflight or İnav but ardupilot not work. Default serial protocol value 2. I am changed to 23 but not working. What should serial options value be? Default value 0. My wiring diagram same as picture 
MAMBA_Basic_F405_MK3_F50

**Comment by FairyTalesl** [comment]
My receiver frysky r9-sx and using sbus protocol

**Comment by tenten8401** [comment]
@FairyTalesI I think I may have screwed something up, gimme a bit and I can get you a new hex file to flash.

**Comment by tenten8401** [comment]
@FairyTalesl can you give this hex a try? I set NODMA up on UART1 how it was supposed to be originally I think.
[arducopter_with_bl.zip](https://github.com/ArduPilot/ardupilot/files/6811956/arducopter_with_bl.zip)

**Comment by tridge** [comment]
> What are your thoughts on the 64kb reserved flash? I notice storage size is 16k and bootloader is 16k but we have 64k reserved..

you could drop it to 48k. We need 2 sectors for flash based storage, so minimum is 3 sectors. By dropping to 3 sectors you will get 16k more for features.
The older boards use 64k as they copied a layout that was done for an early HAL that used one sector for OSD data. We don't do that in HAL_ChibiOS, so it is wasted space, but we don't change it as it would be incompatible with existing bootloaders

**Comment by FairyTalesl** [comment]
Unfortunately, the situation is the same, nothing has changed. Sorry for the late reply, I can't spare time because I'm busy.

**Comment by FairyTalesl** [comment]
UART2 worked directly. But UART1 doesn't work :))

**Comment by tenten8401** [comment]
@FairyTalesl does UART1 work when you set BRD_ALT_CONFIG=1?

**Comment by FairyTalesl** [comment]
Not working. I thought UART2 was working, but when I move sticks, the bars seem to move according to the mission planner's head. I'm thinking of buying a new fc support ardupilot

**Comment by tenten8401** [comment]
@FairyTalesl have you tried using uninverted F.port (soldering to the little pad on the rx)? I use CRSF and it's working great, I don't think a new FC will help.

**Comment by FairyTalesl** [comment]
I am trying SBUS

**Comment by tenten8401** [comment]
I'll see if I can get sbus working on my receiver shortly. It has sbus output.

**Comment by FairyTalesl** [comment]
I finally solved the problem. The problem is caused by the pin configuration you have made, I rolled up my sleeves and installed ubuntu and recompiled it with the correct settings and it worked without any settings. But the osd doesn't work, I'm working on it, is it working for you now? 
Correct file https://github.com/FairyTalesl/ardupilot/blob/master/libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat

**Comment by tenten8401** [comment]
I'm using DJI FPV now, I guess I haven't tested the normal analog OSD :(
Let me see if I can plug in some analog gear and check what I get, in theory it should work as the pins for that were copied direct from betaflight.

**Comment by tenten8401** [comment]
@FairyTalesl what changed in your file? did you forget to push to github? Diffchecker says our files are identical.. I wonder if I screwed up compiling the hex I sent you or something

**Comment by FairyTalesl** [comment]
Sorry check now

**Comment by tenten8401** [comment]
Ooh I see what I screwed up now, don't know how that made it past. Thanks for figuring that out :)
I will look into the analog OSD shortly, do you mind if I add your work to my PR?

**Comment by FairyTalesl** [comment]
No problem let's solve the osd please I want to fly first :))

**Comment by tenten8401** [comment]
Just to confirm before I start checking, OSD_TYPE = 1 for you correct? @FairyTalesl

**Comment by tenten8401** [comment]
I feel the OSD may have something to do with this line:
https://github.com/ArduPilot/ardupilot/blob/32881ba14d823db82731cb50d5dd2d3dde13cdea/libraries/AP_HAL_ChibiOS/hwdef/MambaF405US-I2C/hwdef.dat#L117

Everything else matches Betaflight target. Could you try changing DEVID maybe? DEVID matches MambaF405v2 but it may not be the same.

**Comment by FairyTalesl** [comment]
Yes it is set to 1. I think the font file is missing.

https://github.com/FairyTalesl/ardupilot/blob/db82adbb149fc89d2290e7b36b9cd61ccd07c862/libraries/AP_HAL_ChibiOS/hwdef/MambaF405v2/hwdef.dat#L153

**Comment by tenten8401** [comment]
@FairyTalesl oops, you are correct. I removed it as part of tridge's comment saying it was automatic now, but looking back I think he was only referring to the STM32 advanced pins.
image
Can you test with the font reinstalled please?

**Comment by FairyTalesl** [comment]
Okay  i will test it and get back

**Comment by FairyTalesl** [comment]
Yess working baby :))

**Comment by tenten8401** [comment]
@FairyTalesl Awesome. I will merge your fixes into my pull request.
If you are planning on flying long term with this it wouldn't hurt to compile your hex with beta code instead of master.

**Comment by FairyTalesl** [comment]
Ok, now I'm going to mess with some adjustments. I made the first flight, I'm testing it like it's okay

**Comment by FairyTalesl** [comment]
UART5 FPORT working now. But it doesn't work without battery. It didn't occur to me to put a battery in previous attempts. The battery was powering the inverter circuit but not getting power from the USB. 👍

**Review by tridge — APPROVED** [review]

**Comment by tenten8401** [comment]
@FairyTalesl Looks like this one got merged early, I'm about to push all those changes to another PR and you will need to re-flash (& update bootloader because 48k for more features) so use iNav configurator or something to flash the .hex with bl file.

**Comment by Hwurzburg** [comment]
@tenten8401 I am working on adding this information into the wiki...the product links shown are out of date....and I am not sure that the lite version even exists any more...please give me the updated links and check that the fc still adheres to the this hwdef target because the lite vesion has a different betaflight target advertised now...thanks

**Comment by tenten8401** [comment]
@Hwurzburg Looks like Diatone updated their site, they were working when I submitted it :(
I can't seem to find it on their website either now, which is strange because they are very much still selling them and restocking them with retailers. It may just be best to link to GetFPV's product pages for them:
https://www.getfpv.com/diatone-mamba-basic-stack-f405-mini-mk3-5-fc-f40-40a-blheli-s-mini-esc.html
https://www.getfpv.com/diatone-mamba-basic-stack-f405-mk3-lite-fc-f40-40a-blheli-s-esc.html

**Comment by tenten8401** [comment]
Hey, you can find the hex in MambaF405US-I2C

Feb 12, 2022 12:12:00 PM Toni Vucic ***@***.***>:

**Pull request merged**